tests_e2e/tests/agent_wait_for_cloud_init/agent_wait_for_cloud_init.py

```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python3
+
+# Microsoft Azure Linux Agent
+#
+# Copyright 2018 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import time
+
+from assertpy import fail
+
+from tests_e2e.tests.lib.agent_test import AgentVmTest
+from tests_e2e.tests.lib.logging import log
+from tests_e2e.tests.lib.shell import CommandError
+from tests_e2e.tests.lib.ssh_client import SshClient
+
+
+class AgentWaitForCloudInit(AgentVmTest):
+    """
+    This test verifies that the Agent waits for cloud-init to complete before it starts processing extensions.
+
+    To do this, it adds 'CloudInitScript' in cloud-init's custom data. The script ensures first that the Agent
+    is waiting for cloud-init, and then sleeps for a couple of minutes before completing. The scripts appends
+    a set of known messages to waagent.log, and the test simply verifies that the messages are present in the
+    log in the expected order, and that they occur before the Agent reports that it is processing extensions.
+    """
+    CloudInitScript = """#!/usr/bin/env bash
+        set -euox pipefail
+    
+        echo ">>> $(date) cloud-init script begin" >> /var/log/waagent.log
+        while ! grep 'Waiting for cloud-init to complete' /var/log/waagent.log; do
+            sleep 15
+        done
+        echo ">>> $(date) The Agent is waiting for cloud-init, will pause for a couple of minutes" >> /var/log/waagent.log
+        sleep 120
+        echo ">>> $(date) cloud-init script end" >> /var/log/waagent.log
+    """
+
+    def run(self):
+        ssh_client: SshClient = self._context.create_ssh_client()
+
+        log.info("Waiting for Agent to start processing extensions")
+        for _ in range(15):
+            try:
+                ssh_client.run_command("grep 'ProcessExtensionsGoalState started' /var/log/waagent.log")
```

Comment:
we have retry function, can we use that?

Reply:
retry_if_false is too simplistic. I would need to write more code than this simple loop to make it work for the specific errors I am handling, with the specific error messages I am logging. It is not worth it.